theano/gpuarray/opt.py

```diff
@@ -210,6 +210,52 @@ def apply(self, fgraph):
                     0, 'fast_run', 'fast_compile', 'merge')
 
 
+class GraphToGPU(Optimizer):
+    """
+    Transfer the graph as a whole to GPU instead of replacing nodes
+    """
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(toolbox.ReplaceValidate())
+
+    def apply(self, fgraph):
+        mapping = {}
+        move_to_GPU = True
+
+        # Building a new graph
+        for i in fgraph.inputs:
+            mapping[i] = GpuFromHost(None)(i)
```

Comment:
> only do that is if i is of type TensorVariable.

Also, I think it is not correct to add only if they're of only `TensorVariable` type. From the test cases, I see `CudaNdarrayType`, `GpuArrayType` and `TensorConstant` being used.  Is it fine if I make a generic check whether it is of the type `Type` or `Variable` ?